cli/src/commands/explore/mod.rs

```diff
@@ -210,11 +219,38 @@ pub mod tests {
             "Encoded call data:\n    0x24040400020cffffff00\n"
         );
         // check that we can explore a certain constant:
-        let output = simulate_run("Balances constants ExistentialDeposit").await;
+        let output = run_against_file("Balances constants ExistentialDeposit").await;
         assert_eq!(output.unwrap(), "Description:\n    The minimum amount required to keep an account open. MUST BE GREATER THAN ZERO!\n\nThe constant has the following shape:\n    u128\n\nThe value of the constant is:\n    33333333\n");
         // check that we can explore a certain storage entry:
-        let output = simulate_run("System storage Account").await;
+        let output = run_against_file("System storage Account").await;
         assert!(output.unwrap().starts_with("Usage:\n    subxt explore System storage Account <KEY_VALUE>\n\nDescription:\n    The full account information for a particular account ID."));
         // in the future we could also integrate with substrate-testrunner to spawn up a node and send an actual storage query to it: e.g. `subxt explore System storage Digest`
     }
+
+    #[tokio::test]
+    async fn insecure_urls_get_denied() {
+        // connection should work fine:
+        run("--url wss://rpc.polkadot.io:443").await.unwrap();
+
+        // errors because the --allow-insecure is not set:
+        assert!(run("--url ws://rpc.polkadot.io:443")
+            .await
+            .unwrap_err()
+            .to_string()
+            .contains("--allow-insecure"));
+
+        // also errors, but later, because of networking reasons:
+        assert!(run("--url ws://rpc.polkadot.io:443 --allow-insecure")
+            .await
+            .unwrap_err()
+            .to_string()
+            .contains("Request error"));
+
+        // also errors, but later, because localhost is considered secure, even without the flag
+        assert!(run("--url ws://localhost")
+            .await
+            .unwrap_err()
+            .to_string()
+            .contains("Request error"));
```

Comment:
Would be good to assert that the error indicates that url in insecure and run with `--allow-insecure` to fix that.

IIRC, you added a descriptive error and just change the assert here should do it.

Reply:
The point of the test against `--url ws://localhost` is that it should be accepted by default, even though the scheme is insecure. That is why we expect a "Request error" here instead of an error that prompts the user to use "--allow-insecure".

I know the string checking is not super nice, but this is an end to end test in the CLI tool, which just returns a color_eyre::Result<String>, so I lose the exact error type.